src/nxt_array.c

```diff
@@ -1,150 +1,253 @@
+static nxt_uint_t nxt_array_calc_alloc(nxt_uint_t nelts);
+static void *nxt_array_elts_realloc(nxt_array_t *array,
+                                    nxt_uint_t nalloc);
+
+
+/*
+ * nxt_array_create - Create a new array and return a pointer to it.
+ */
 nxt_array_t *
-nxt_array_create(nxt_mp_t *mp, nxt_uint_t n, size_t size)
+nxt_array_create(nxt_mp_t *mp, nxt_uint_t arr_size, size_t elt_size)
 {
-    nxt_array_t  *array;
+    nxt_uint_t     nalloc;
+    nxt_array_t    *array;
+    void           *elts;
 
-    array = nxt_mp_alloc(mp, sizeof(nxt_array_t) + n * size);
+    array = nxt_mp_alloc(mp, sizeof(nxt_array_t));
 
     if (nxt_slow_path(array == NULL)) {
         return NULL;
     }
 
-    array->elts = nxt_pointer_to(array, sizeof(nxt_array_t));
+    nalloc = nxt_array_calc_alloc(arr_size);
+    elts = nxt_mp_alloc(mp, elt_size * nalloc);
+
+    if (nxt_slow_path(elts == NULL)) {
+        return NULL;
+    }
+
+    array->elts = elts;
     array->nelts = 0;
-    array->size = size;
-    array->nalloc = n;
+    array->size = elt_size;
+    array->nalloc = nalloc;
     array->mem_pool = mp;
 
     return array;
 }
 
 
+/*
+ * nxt_array_destroy - Destroy the existing array.
+ */
 void
 nxt_array_destroy(nxt_array_t *array)
 {
-    if (array->elts != nxt_pointer_to(array, sizeof(nxt_array_t))) {
-        nxt_mp_free(array->mem_pool, array->elts);
-    }
+    if (nxt_fast_path(array != NULL)) {
+
+        array->nelts = 0;
 
-    nxt_mp_free(array->mem_pool, array);
+        (void) nxt_mp_free(array->mem_pool, array->elts);
+        (void) nxt_mp_free(array->mem_pool, array);
+
+    }
 }
 
 
+/*
+ * nxt_array_add - Add an element to the existing array.
+ */
 void *
 nxt_array_add(nxt_array_t *array)
 {
-    void      *p;
-    uint32_t  nalloc, new_alloc;
-
-    nalloc = array->nalloc;
+    if (nxt_fast_path(array != NULL)) {
 
-    if (array->nelts == nalloc) {
+        nxt_uint_t    nalloc;
+        void          *elt, *elts;
 
-        if (nalloc < 16) {
-            /* Allocate new array twice larger than current. */
-            new_alloc = (nalloc == 0) ? 4 : nalloc * 2;
+        if (array->nelts == array->nalloc) {
 
-        } else {
-            /* Allocate new array 1.5 times larger than current. */
-            new_alloc = nalloc + nalloc / 2;
-        }
+            nalloc = nxt_array_calc_alloc(array->nelts);
+            elts = nxt_array_elts_realloc(array, nalloc);
 
-        p = nxt_mp_alloc(array->mem_pool, array->size * new_alloc);
+            if (nxt_slow_path(elts == NULL)) {
+                return NULL;
+            }
 
-        if (nxt_slow_path(p == NULL)) {
-            return NULL;
         }
 
-        nxt_memcpy(p, array->elts, array->size * nalloc);
+        elt = nxt_pointer_to(array->elts, array->size * array->nelts);
 
-        if (array->elts != nxt_pointer_to(array, sizeof(nxt_array_t))) {
-            nxt_mp_free(array->mem_pool, array->elts);
```

Comment:
- eliminated the possibility of early deletion of an existing pointer to a set of elements (elts) at the time of memory reallocation;

`array->elts` is freed before the new pointer is set.

Reply:
There's nothing wrong in freeing `array->elts` pointer before setting another one. No parallel access is expected to arrays anyway.

Reply:
The fact that parallel access to arrays is not supposed to be said somewhere? Maybe in other not trivial and not silly comments?

**Question to you from the opposite:**
What's wrong with assigning a new pointer before deleting the old pointer?

Reply:
Have you seen any locking in the array functions? For example, nxt_array_add() is clearly not atomic with or without your patch. Actually, it's ok to assume that there's no parallel access supposed to any of the Unit's changeable data structures unless they have some locking around. The memory pools and every code, that uses memory pools isn't atomic too. We try to avoid parallel access as much as possible, keeping such structures and places at minimum among the project.

Reply:
> What's wrong with assigning a new pointer before deleting the old pointer?

Sure there's nothing wrong. Except it just adds a bit more code, a bit more logic that needs to be reviewed and will be read by everybody. Much worse if seeing such handling of previous pointer they will make some assumption about the reasons why it was made like that.

Any variant of code makes sense and would look fine in this particular case, but there should be a reason for changing already existing code. Current code is dumb and straightforward right as is as it's required in this particular place.

Reply:
I mean, is there anything to warn other developers against using array-based memory manipulation in parallel, other than internal rules known and available only to your team?

As you aptly noted, the operation of adding an element to the array was not atomic either before or after the changes I proposed, as was the parallel use, so using this argument as an excuse for not adjusting, albeit insignificant, but nevertheless not useless, looks extremely strange .

**You still haven't answered the question:**
>
> **Question to you from the opposite:**
> What's wrong with assigning a new pointer before deleting the old pointer?

Reply:
> I mean, is there anything to warn other developers against using array-based memory manipulation in parallel, other than internal rules known and available only to your team?

I'd rather warn any developers of using any function without reading and understanding the function's code. And it's just a common sense especially when dealing with some internal APIs of the project.

Even public APIs of some very popular projects are very far from being well-understood without reading the internals. I can't count how many times I had to read the Linux kernel's code behind some syscall to understand nuances of its functioning that not covered anywhere. And for sure, anyone will fail, trying to use PHP's interpreter API or OpenSSL only by manuals and without reading internals.

So, the answer to your question is experience. It takes some time to get familiar with the project's architecture and the right way how to code it. And a change to how the particular pointer is manipulated won't make this code thread safe.

> What's wrong with assigning a new pointer before deleting the old pointer?

I don't know if while writing it you've seen my [second message](https://github.com/nginx/unit/pull/664/commits/8d1df4121295f929c7b2b5ea974f223ca25e50d5#r836395146) to the thread or not, but anyway, in short:

1. Making non thread safe code to look in some parts like it can be thread safe - only misleads.
2. Nothing wrong, except it wasn't done this way initially for the sake of simplicity and it's not required.

Reply:
I'm sorry, when I answered in the previous message, the page with your answer has not yet been updated. Heard you.